packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,50 @@
+import * as cdk from '@aws-cdk/core';
+import { TagUpdateConstraintOptions } from '../constraints';
+import { IPortfolio } from '../portfolio';
+import { IProduct } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(
+    portfolio: IPortfolio, product: IProduct,
+  ): { associationKey: string, cfnPortfolioProductAssociation: CfnPortfolioProductAssociation } {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, portfolio.stack.node.addr);
+    const constructId = `PortfolioProductAssociation${associationKey}`;
+    const associationNode = portfolio.node.tryFindChild(constructId);
+
+    return {
+      associationKey: associationKey,
+      cfnPortfolioProductAssociation: !associationNode ? new CfnPortfolioProductAssociation(portfolio as unknown as cdk.Resource, constructId, {
```

Comment:
This is way overcomplicated in my opinion.

```ts
  public static associateProductWithPortfolio(
    portfolio: IPortfolio, product: IProduct,
  ): { associationKey: string, cfnPortfolioProductAssociation: CfnPortfolioProductAssociation } {
    const associationKey = hashValues(portfolio.node.addr, product.node.addr, portfolio.stack.node.addr);
    const constructId = `PortfolioProductAssociation${associationKey}`;
    const existingAssociation = portfolio.node.tryFindChild(constructId);
    const cfnAssociation = existingAssociation
      ? existingAssociation as CfnPortfolioProductAssociation
      : new CfnPortfolioProductAssociation(portfolio as unknown as cdk.Resource, constructId, {
        portfolioId: portfolio.portfolioId,
        productId: product.productId,
      });

    return {
      associationKey: associationKey,
      cfnPortfolioProductAssociation: cfnAssociation,
    };
  }
```

Reply:
not sure why didn't try this...  thanks.